spec/index.bs

```diff
@@ -2428,114 +2428,115 @@ A {{GPUBindGroupLayout}} object has the following internal slots:
-        1. If |bindingDescriptor|.{{GPUBindGroupLayoutEntry/type}} is {{GPUBindingType/readonly-storage-texture}} or {{GPUBindingType/writeonly-storage-texture}}:
-            1. Ensure [=storage texture validation=] is not violated.
-            1. Ensure [=storage texture format validation=] is not violated.
-        1. If |bindingDescriptor|.{{GPUBindGroupLayoutEntry/type}} is {{GPUBindingType/sampler}} or {{GPUBindingType/comparison-sampler}}
-            , ensure [=sampler validation=] is not violated.
-        1. Insert |bindingDescriptor| into |layout|.{{GPUBindGroupLayout/[[entryMap]]}}
-            with the key of |bindingDescriptor|/{{GPUBindGroupLayoutEntry/binding}}.
-    1. Return |layout|.
-
-    <div class=validusage dfn-for=GPUDevice.createBindGroupLayout>
-        <dfn abstract-op>Valid Usage</dfn>
-
-        If any of the following conditions are violated:
-            1. Generate a {{GPUValidationError}} in the current scope with appropriate error message.
-            1. Create a new [=invalid=] {{GPUBindGroupLayout}} and return the result.
-
-        <dfn>bind group layout device validation</dfn>: The {{GPUDevice}} must not be lost.
-
-        <dfn>binding validation</dfn>: Each |bindingDescriptor|.{{GPUBindGroupLayoutEntry/binding}} in |descriptor| must be unique.
-
-        <dfn>vertex shader binding validation</dfn>: {{GPUBindingType/storage-buffer}} and {{GPUBindingType/writeonly-storage-texture}} are not allowed.
+<dl dfn-type=method dfn-for=GPUDevice>
+    : <dfn>createBindGroupLayout(descriptor)</dfn>
+    ::
 
-        <dfn>uniform buffer validation</dfn>: There must be {{GPULimits/maxUniformBuffersPerShaderStage|GPULimits.maxUniformBuffersPerShaderStage}} or
-            fewer |bindingDescriptor|s of type {{GPUBindingType/uniform-buffer}} visible on each shader stage in |descriptor|.
+    Creates a {{GPUBindGroupLayout}}.
 
-        <dfn>dynamic uniform buffer validation</dfn>: There must be {{GPULimits/maxDynamicUniformBuffersPerPipelineLayout|GPULimits.maxDynamicUniformBuffersPerPipelineLayout}} or
-                fewer |bindingDescriptor|s of type {{GPUBindingType/uniform-buffer}} with {{GPUBindGroupLayoutEntry/hasDynamicOffset}} set to `true` in
-                |descriptor| that are visible to any shader stage.
+    <div algorithm=GPUDevice.createBindGroupLayout>
+        **Called on:** {{GPUDevice}} |this|.
 
-        <dfn>storage buffer validation</dfn>: There must be {{GPULimits/maxStorageBuffersPerShaderStage|GPULimits.maxStorageBuffersPerShaderStage}} or
-            fewer |bindingDescriptor|s of type {{GPUBindingType/storage-buffer}} visible on each shader stage in |descriptor|.
+        **Arguments:**
+        <pre class=argumentdef for="GPUDevice/createBindGroupLayout(descriptor)">
+            |descriptor|:
+        </pre>
+
+        **Returns:** {{GPUBindGroupLayout}}
+
```

Comment:
The whole algorithm should be on the device timeline. Optional if you want to figure out how to fix this now. (Have to do something like create the object, then issue the device timeline steps.)

Reply:
Happy to try and get it in this PR, but I think we're going to need to iterate on it a bit. Not sure how to write out the algorithm across the different timelines in a synchronous function. (My text about "returning" from the device timeline is almost certainly wrong.) Is the intent that the object returns immediately and then may get marked invalid asynchronously at a later point?

Reply:
Essentially, yes. The only things that will check whether the device is valid should be on the device timeline. I was intending to try to figure out exactly how to explicitly separate device and client state (with the internal objects) but I haven't been able to get to it. Might end up creating a concept of a handle? maybe something like:

- create a handle
- issue steps on device timeline:
    - do stuff and make handle refer to a valid or invalid internal object
- return GPUBindGroupLayout(handle)